formulaic/materializers/narwhals.py

```diff
@@ -0,0 +1,248 @@
+from __future__ import annotations
+
+import functools
+import itertools
+from typing import TYPE_CHECKING, Any, Dict, List, Sequence, Set, Tuple, cast
+
+import narwhals.stable.v1 as nw
+import numpy
+import pandas
+import scipy.sparse as spsparse
+from interface_meta import override
+
+# Prefer these for `isinstance` checks so they catch both
+# narwhals.stable.v1.DataFrame and narwhals.DataFrame
+from narwhals import DataFrame as NwDataFrame
+from narwhals import Series as NwSeries
```

Comment:
I don't like this, we should address it in Narwhals - I've made an issue (https://github.com/narwhals-dev/narwhals/issues/1548), will update once that's addressed / released

Reply:
done